docs/JavaCodePerformance.md

```diff
@@ -2511,12 +2511,62 @@ class SingletonBMGood {
 }
 ```
 
+Improper program flow
+---------------------
+
+#### IPF01
+
+**Observation: A method calls itself, also known as recursion, and it doesn't have a proper and guaranteed stop condition.**  
+**Problem:** It may become an infinite loop and result in an OutOfMemoryError or StackOverflowError, and high CPU usage.  
+**Solution:** limit the number of recursive calls: use a counter to count up-to or down-from a maximum number of calls, for every recursive call, and stop when the maximum is reached.   
+**Rule name:** AvoidInfiniteRecursion.  
+**Example:**
+```java
+class InfiniteRecursionBad {
+  private void foo() {
+    boolean success = tryRemoteCall();
+      if (!success) {
+        delay(10, MILLISECONDS);
+        foo(); // bad
+    }
+  }
+}
+
+class FiniteRecursionGood {
+  private static final int MAX_ATTEMPTS = 5;
+
+  private void foo(int attemptsLeft) {
+    boolean success = tryRemoteCall();
+    if (!success && attemptsLeft > 0) {
+      delay(10, MILLISECONDS);
+      foo(--attemptsLeft);
+    }
+  }
+}
+
+// when called from multiple threads, shared as a singleton, it needs to be thread-safe
+class FiniteRecursionSingletonThreadSafeGood {
+  private static final int MAX_ATTEMPTS = 5;
+  private final AtomicInteger attemptsLeft = new AtomicInteger(MAX_ATTEMPTS); // thread safe
+
+  private void foo() {
+    boolean success = tryRemoteCall();
+    if (!success && attemptsLeft.decrementAndGet() >= 0 ) {
+      delay(10, MILLISECONDS);
+      foo();
+    }
+  }
+}
+```
+**See:** [Java recursion by Baeldung](https://www.baeldung.com/java-recursion)
```

Comment:
do we have a link with less commercials? :-) Maybe this one: https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/examples-Java-recursion-recursive-methods

This also advices to not use recursion in production code. Can this retry example also be done without recursion? That would be better solution. Max attempts with retries is a good idea and also supported by @Retry annotations. But max calls in a calculation like Fibonacci algos? What number would you use?

Maybe better advice to avoid recursion if possible (until Java solves this or you use alternatives that are not really recursive in call stack nature: https://stackoverflow.com/questions/860550/stack-overflows-from-deep-recursion-in-java). So "AvoidRecursion"?

Reply:
Note that the rule can also be applied to loops: is there in all cases an escape from the loop? (Can you make a rule for that?)

Reply:
I'don't accept unnecessary cookies, so I don't have commercials. I can add that one. 

Max attempts for algos like Fibonacci should be a large enough number to accommodate the calculations it is used for, and it should be tested.

A rule to AvoidRecursion may be a good idea, then on a lower severity level.  
A rule for a loop would also be useful.

Reply:
I will add to the description that recursion may not be a good idea in general.